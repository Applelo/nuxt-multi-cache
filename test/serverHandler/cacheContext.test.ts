import { beforeEach, beforeAll, describe, expect, test, vi } from 'vitest'
import cacheContextHandler from './../../src/runtime/serverHandler/cacheContext'

vi.mock('#imports', () => {
  return {
    useRuntimeConfig: () => {
      return {
        multiCache: {},
      }
    },
  }
})

vi.mock('@nuxt/kit', () => {
  return {
    loadNuxtConfig: () => {
      return Promise.resolve({
        multiCache: {
          component: {
            enabled: true,
          },
          enabledForRequest: (event: any) => {
            return Promise.resolve(!!event.__ENABLED_FOR_REQUEST)
          },
        },
      })
    },
  }
})

describe('cacheContext server handler', () => {
  beforeEach(() => {
    vi.resetAllMocks()
    vi.resetModules()
  })
  beforeAll(() => {
    vi.resetModules()
  })
  test('Adds cache context to event', async () => {
    const event: any = {
      context: {},
      __ENABLED_FOR_REQUEST: true,
    }
    await cacheContextHandler(event)
    expect(event.context).toMatchSnapshot()
  })

  test('Does not add cache context to event if enabledForRequest returns false', async () => {
    const event: any = {
      context: {},
      __ENABLED_FOR_REQUEST: false,
    }
    await cacheContextHandler(event)
    expect(event.context).toEqual({})
  })

  test('Caches that there is no enabledForRequest method.', async () => {
    const event: any = {
      context: {},
    }
    const shouldAddCacheContext = await import(
      './../../src/runtime/serverHandler/cacheContext'
    ).then((v) => v.shouldAddCacheContext)

    const kit = await import('@nuxt/kit')
    kit.loadNuxtConfig = vi.fn().mockReturnValueOnce(
      Promise.resolve({
        multiCache: {},
      }),
    )
    expect(await shouldAddCacheContext(event)).toEqual(true)
    expect(await shouldAddCacheContext(event)).toEqual(true)
  })
})
