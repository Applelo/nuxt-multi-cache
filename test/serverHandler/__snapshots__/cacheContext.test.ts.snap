// Vitest Snapshot v1

exports[`cacheContext server handler > Adds cache context to event 1`] = `
{
  "__MULTI_CACHE": {
    "component": {
      "clear": [Function],
      "dispose": [Function],
      "getItem": [Function],
      "getKeys": [Function],
      "getMeta": [Function],
      "hasItem": [Function],
      "mount": [Function],
      "removeItem": [Function],
      "removeMeta": [Function],
      "setItem": [Function],
      "setMeta": [Function],
      "unmount": [Function],
      "unwatch": [Function],
      "watch": [Function],
    },
  },
  "__MULTI_CACHE_ROUTE": {
    "cacheable": null,
    "control": CacheControl {
      "immutable": null,
      "maxAge": null,
      "maxStale": null,
      "maxStaleDuration": null,
      "minFresh": null,
      "mustRevalidate": null,
      "noCache": null,
      "noStore": null,
      "noTransform": null,
      "onlyIfCached": null,
      "private": null,
      "proxyRevalidate": null,
      "public": null,
      "sharedMaxAge": null,
      "staleIfError": null,
      "staleWhileRevalidate": null,
    },
    "tags": [],
  },
}
`;
